Mathlib/CategoryTheory/MarkovCategory/Basic.lean

```diff
@@ -0,0 +1,281 @@
+lemma copy_natural_of_deterministic {X Y : C} (f : X ⟶ Y)
+    (h : f ≫ Δ[Y] = Δ[X] ≫ (f ⊗ₘ f)) : f ≫ Δ[Y] = Δ[X] ≫ (f ⊗ₘ f) := h
+
+/-- The identity morphism preserves the comonoid structure -/
+@[simp]
+lemma id_copy (X : C) : 𝟙 X ≫ Δ[X] = Δ[X] := Category.id_comp _
+
+/-- The identity morphism preserves the delete operation -/
+@[simp]
+lemma id_del (X : C) : 𝟙 X ≫ ε[X] = ε[X] := Category.id_comp _
+
+/-- **Left counit law**: Copying then deleting the first copy recovers the original.
+This is a fundamental property of the comonoid structure, ensuring that marginalization
+is the inverse of creating correlation. -/
+@[simp, reassoc]
+lemma copy_comp_del_left (X : C) : Δ[X] ≫ (ε[X] ▷ X) = (λ_ X).inv := counit_comul X
+
+/-- **Right counit law**: Copying then deleting the second copy recovers the original.
+Together with `copy_comp_del_left`, this ensures both marginals equal the original
+distribution. -/
+@[simp, reassoc]
+lemma copy_comp_del_right (X : C) : Δ[X] ≫ (X ◁ ε[X]) = (ρ_ X).inv := comul_counit X
+
+/-- Copying is commutative: swapping the copies gives the same result.
+This shows that correlation is symmetric. -/
+@[simp, reassoc]
+lemma copy_comp_braiding (X : C) : Δ[X] ≫ (β_ X X).hom = Δ[X] := copy_comm X
+
+/-- **Coassociativity**: Copying then copying the first output equals copying then
+copying the second output (up to reassociation). This means we can create multiple
+correlated copies without ambiguity. -/
+@[simp, reassoc]
+lemma copy_comp_copy_left (X : C) :
+    Δ[X] ≫ (Δ[X] ▷ X) = Δ[X] ≫ (X ◁ Δ[X]) ≫ (α_ X X X).inv := coassoc X
+
+/-- Copying distributes over tensor products via the middle four interchange.
+This shows how copying preserves the independence structure. -/
+@[simp]
+lemma copy_tensor_eq (X Y : C) :
+    Δ[X ⊗ Y] = (Δ[X] ⊗ₘ Δ[Y]) ≫ middleFourInterchange X X Y Y := copy_tensor X Y
+
+/-- Deletion distributes over tensor products.
+Marginalizing over a product equals marginalizing over each component. -/
+@[simp]
+lemma del_tensor_eq (X Y : C) :
+    ε[X ⊗ Y] = (ε[X] ⊗ₘ ε[Y]) ≫ (λ_ (𝟙_ C)).hom := del_tensor X Y
+
+/-- Deletion is natural: it commutes with all morphisms.
+This forces the monoidal unit to be terminal. -/
+@[simp, reassoc]
+lemma del_natural_simp {X Y : C} (f : X ⟶ Y) : f ≫ ε[Y] = ε[X] := del_natural f
+
+end MarkovCategory
+
+/-- A morphism in a Markov category is deterministic if it preserves the copy operation.
+
+Deterministic morphisms are the "pure" functions that preserve correlations perfectly.
+In truly probabilistic categories, these are the non-randomized morphisms.
+All morphisms in cartesian categories are deterministic, but a Markov category
+can have both deterministic and truly stochastic morphisms. -/
+class Deterministic {C : Type u} [Category.{v} C] [MonoidalCategory.{v} C] [MarkovCategory C]
+{X Y : C} (f : X ⟶ Y) : Prop where
+  preserves_copy : f ≫ MarkovCategory.copyMor Y = MarkovCategory.copyMor X ≫ (f ⊗ₘ f)
```

Comment:
This should be made an abbrev for a `IsComon_Hom` instance. This should make the two instances below automatic.